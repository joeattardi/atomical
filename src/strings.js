export const months = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
];

export const fullDayNames = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
];

export const shortDayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

export const initialDayNames = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];
