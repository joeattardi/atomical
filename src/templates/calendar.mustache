<div class="{{classes.calendar}}">
  <div class="{{classes.header}}">
    <button class="{{classes.previousButton}}">{{{icons.previous}}}</button>
    <h2><span class="{{classes.titleMonth}}">{{month}}</span> <span class="{{classes.titleYear}}">{{year}}</span></h2>
    <button class="{{classes.nextButton}}">{{{icons.next}}}</button>
  </div>
  <div class="{{classes.days}}">
    {{#dayLabels}}<div class="{{classes.dayHeader}}">{{.}}</div>{{/dayLabels}}
  </div>
  <div class="{{classes.grid}}">
    {{#beginPlaceholders}}
      <div class="{{classes.placeholder}}"></div>
    {{/beginPlaceholders}}

    {{#days}}
      <div class="{{classes}}"><div>{{date}}</div></div>
    {{/days}}

    {{#endPlaceholders}}
      <div class="{{classes.placeholder}}"></div>
    {{/endPlaceholders}}
  </div>
</div>
