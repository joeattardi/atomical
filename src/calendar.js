import styles from '../css/atomical.css';

import {
  months,
  fullDayNames,
  shortDayNames,
  initialDayNames
} from './strings';
import { createMainTemplate } from './templates';
import {
  empty,
  getSizeClass,
  getDateFormat,
  createDay,
  createEvent,
  createPlaceholder,
  increment
} from './util';

const mainTemplate = createMainTemplate(
  fullDayNames,
  shortDayNames,
  initialDayNames
);

export class Calendar extends HTMLElement {
  constructor() {
    super();
    this.eventDays = {};
  }

  connectedCallback() {
    const styleEl = document.createElement('style');
    styleEl.textContent = styles;

    const shadow = this.attachShadow({ mode: 'open' });
    shadow.appendChild(styleEl);
    shadow.appendChild(mainTemplate.content.cloneNode(true));

    this.findElements();
    this.render();

    this.resizeObserver = new ResizeObserver(entries => {
      const container = entries.find(entry => entry.target === this.calendarEl);
      const { width } = container.contentRect;
      this.calendarEl.className = `calendar ${getSizeClass(width)}`;
    });
    this.resizeObserver.observe(this.calendarEl);
  }

  render() {
    const monthAttr = this.getAttribute('month');
    const yearAttr = this.getAttribute('year');

    const now = new Date();
    const month = monthAttr !== null ? parseInt(monthAttr, 10) : now.getMonth();
    const year = yearAttr !== null ? parseInt(yearAttr, 10) : now.getFullYear();

    const firstDay = new Date(year, month, 1);
    const lastDay = new Date(year, month + 1, 0);
    const daysInMonth = lastDay.getDate();
    const firstWeekday = firstDay.getDay();
    const lastWeekday = lastDay.getDay();

    const leadingPlaceholderCount = firstWeekday;
    const trailingPlaceholderCount = fullDayNames.length - 1 - lastWeekday;

    this.monthEl.textContent = months[month];
    this.yearEl.textContent = year;

    empty(this.gridEl);

    let row = 1;
    let column = 1;

    for (let i = 0; i < leadingPlaceholderCount; i++) {
      this.gridEl.appendChild(createPlaceholder(row, column));
      [row, column] = increment(row, column);
    }

    for (let i = 1; i <= daysInMonth; i++) {
      this.gridEl.appendChild(createDay(i, month, year, row, column));

      const dateFormat = getDateFormat(new Date(year, month, i));
      const eventsInDay = this.eventDays[dateFormat];

      if (eventsInDay && eventsInDay.length) {
        eventsInDay.forEach((event, i) => {
          this.gridEl.appendChild(createEvent(event, i, row, column));
        });
      }

      [row, column] = increment(row, column);
    }

    for (let i = 0; i < trailingPlaceholderCount; i++) {
      this.gridEl.appendChild(createPlaceholder(row, column));
      [row, column] = increment(row, column);
    }

    this.isRendered = true;
  }

  findElements() {
    this.calendarEl = this.shadowRoot.querySelector('.calendar');
    this.monthEl = this.shadowRoot.querySelector('.month');
    this.yearEl = this.shadowRoot.querySelector('.year');
    this.gridEl = this.shadowRoot.querySelector('.grid');
  }

  disconnectedCallback() {
    this.resizeObserver.disconnect();
  }

  attributeChangedCallback() {
    if (this.isRendered) {
      this.render();
    }
  }

  addEvent(event) {
    const dateFormat = getDateFormat(event.start);
    if (!this.eventDays[dateFormat]) {
      this.eventDays[dateFormat] = [];
    }

    this.eventDays[dateFormat].push(event);

    this.render();
  }

  static get observedAttributes() {
    return ['month', 'year'];
  }
}

customElements.define('atomical-calendar', Calendar);
