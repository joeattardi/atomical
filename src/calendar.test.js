import ResizeObserver from 'resize-observer-polyfill';

import { Calendar } from './calendar';
import { months } from './strings';

window.ResizeObserver = ResizeObserver;

describe('Atomical Calendar', () => {
  test('should display the current month and year', () => {
    const calendar = new Calendar();
    document.body.appendChild(calendar);

    const today = new Date();
    expect(calendar.monthEl.textContent).toEqual(months[today.getMonth()]);
    expect(calendar.yearEl.textContent).toEqual(today.getFullYear().toString());
    expect(calendar.shadowRoot.innerHTML).toMatchSnapshot();
  });

  test('should display the specified month and year', () => {
    const calendar = new Calendar();
    calendar.setAttribute('month', 5);
    calendar.setAttribute('year', 2020);

    document.body.appendChild(calendar);

    expect(calendar.monthEl.textContent).toEqual(months[5]);
    expect(calendar.yearEl.textContent).toEqual('2020');
    expect(calendar.shadowRoot.innerHTML).toMatchSnapshot();
  });
});
