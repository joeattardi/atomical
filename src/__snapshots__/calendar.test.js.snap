// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atomical Calendar should display the current month and year 1`] = `
"<style>:host {
  --header-background: #cc333f;
  --header-color: #ffffff;
  --button-hover-color: #1693a5;
  --today-background: #f5e7dc;
  --placeholder-background: #eeeeee;
  --day-border-color: #cccccc;
  --calendar-background: #ffffff;

  font-family: Arial, Helvetica, sans-serif;
  margin: 1rem;
  font-size: 1rem;
}

.header {
  width: 100%;
  text-align: center;
  display: flex;
  flex-direction: row;
  padding: 0;
  padding-top: min(3%, 1rem);
  padding-bottom: min(3%, 1rem);
  justify-content: center;
  color: var(--header-color);
  background: var(--header-background);
  border-radius: 5px 5px 0 0;
  font-size: 1.2em;
}

.medium .header {
  font-size: 1em;
}

.small .header {
  font-size: 0.8em;
}

.header h2 {
  margin: 0;
  flex-grow: 1;
  text-align: center;
}

.header .year {
  font-weight: normal;
}

.header .previousButton,
.header .nextButton {
  background: transparent;
  border: none;
  cursor: pointer;
  transition: all 150ms;
}

.header .previousButton:hover,
.header .nextButton:hover {
  color: var(--button-hover-color);
}

.days {
  display: grid;
  grid-template-rows: 2rem;
  grid-template-columns: repeat(7, 1fr);
  border-bottom: none;
  background: var(--header-background);
  gap: 1px;
}

.daysLarge,
.daysMedium,
.daysSmall {
  display: none;
}

.large .daysLarge,
.medium .daysMedium,
.small .daysSmall {
  display: grid;
}

.days .dayHeader {
  align-self: stretch;
  justify-self: stretch;
  background: var(--header-background);
  color: var(--header-color);
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 0.8em;
}

.grid {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  border: 1px solid var(--day-border-color);
  background: var(--day-border-color);
  gap: 1px;
}

.grid > div {
  display: grid;
  background: var(--calendar-background);
}

.grid > div > div {
  grid-area: 1 / 1 / 2 / 2;
}

.grid > div::before {
  content: '';
  padding-bottom: 100%;
  display: block;
  grid-area: 1 / 1 / 2 / 2;
}

.grid .placeholder {
  background: var(--placeholder-background);
}

.grid .day > div {
  padding: 0.5rem;
}

.small .grid .day > div {
  display: flex;
  align-items: center;
  justify-content: center;
}

.grid .day.today {
  background: var(--today-background);
}
</style>
    <div class=\\"calendar large\\">
      <div class=\\"header\\">
        <h2><span class=\\"month\\">September</span> <span class=\\"year\\">2020</span></h2>
      </div>
      <div class=\\"days daysLarge\\"><div class=\\"dayHeader\\">Sunday</div><div class=\\"dayHeader\\">Monday</div><div class=\\"dayHeader\\">Tuesday</div><div class=\\"dayHeader\\">Wednesday</div><div class=\\"dayHeader\\">Thursday</div><div class=\\"dayHeader\\">Friday</div><div class=\\"dayHeader\\">Saturday</div></div>
      <div class=\\"days daysMedium\\"><div class=\\"dayHeader\\">Sun</div><div class=\\"dayHeader\\">Mon</div><div class=\\"dayHeader\\">Tue</div><div class=\\"dayHeader\\">Wed</div><div class=\\"dayHeader\\">Thu</div><div class=\\"dayHeader\\">Fri</div><div class=\\"dayHeader\\">Sat</div></div>
      <div class=\\"days daysSmall\\"><div class=\\"dayHeader\\">S</div><div class=\\"dayHeader\\">M</div><div class=\\"dayHeader\\">T</div><div class=\\"dayHeader\\">W</div><div class=\\"dayHeader\\">T</div><div class=\\"dayHeader\\">F</div><div class=\\"dayHeader\\">S</div></div>
      <div class=\\"grid\\"><div class=\\"placeholder\\"></div><div class=\\"placeholder\\"></div><div class=\\"day\\"><div>1</div></div><div class=\\"day\\"><div>2</div></div><div class=\\"day\\"><div>3</div></div><div class=\\"day\\"><div>4</div></div><div class=\\"day\\"><div>5</div></div><div class=\\"day\\"><div>6</div></div><div class=\\"day\\"><div>7</div></div><div class=\\"day\\"><div>8</div></div><div class=\\"day\\"><div>9</div></div><div class=\\"day\\"><div>10</div></div><div class=\\"day today\\"><div>11</div></div><div class=\\"day\\"><div>12</div></div><div class=\\"day\\"><div>13</div></div><div class=\\"day\\"><div>14</div></div><div class=\\"day\\"><div>15</div></div><div class=\\"day\\"><div>16</div></div><div class=\\"day\\"><div>17</div></div><div class=\\"day\\"><div>18</div></div><div class=\\"day\\"><div>19</div></div><div class=\\"day\\"><div>20</div></div><div class=\\"day\\"><div>21</div></div><div class=\\"day\\"><div>22</div></div><div class=\\"day\\"><div>23</div></div><div class=\\"day\\"><div>24</div></div><div class=\\"day\\"><div>25</div></div><div class=\\"day\\"><div>26</div></div><div class=\\"day\\"><div>27</div></div><div class=\\"day\\"><div>28</div></div><div class=\\"day\\"><div>29</div></div><div class=\\"day\\"><div>30</div></div><div class=\\"placeholder\\"></div><div class=\\"placeholder\\"></div><div class=\\"placeholder\\"></div></div>
    </div>
  "
`;

exports[`Atomical Calendar should display the specified month and year 1`] = `
"<style>:host {
  --header-background: #cc333f;
  --header-color: #ffffff;
  --button-hover-color: #1693a5;
  --today-background: #f5e7dc;
  --placeholder-background: #eeeeee;
  --day-border-color: #cccccc;
  --calendar-background: #ffffff;

  font-family: Arial, Helvetica, sans-serif;
  margin: 1rem;
  font-size: 1rem;
}

.header {
  width: 100%;
  text-align: center;
  display: flex;
  flex-direction: row;
  padding: 0;
  padding-top: min(3%, 1rem);
  padding-bottom: min(3%, 1rem);
  justify-content: center;
  color: var(--header-color);
  background: var(--header-background);
  border-radius: 5px 5px 0 0;
  font-size: 1.2em;
}

.medium .header {
  font-size: 1em;
}

.small .header {
  font-size: 0.8em;
}

.header h2 {
  margin: 0;
  flex-grow: 1;
  text-align: center;
}

.header .year {
  font-weight: normal;
}

.header .previousButton,
.header .nextButton {
  background: transparent;
  border: none;
  cursor: pointer;
  transition: all 150ms;
}

.header .previousButton:hover,
.header .nextButton:hover {
  color: var(--button-hover-color);
}

.days {
  display: grid;
  grid-template-rows: 2rem;
  grid-template-columns: repeat(7, 1fr);
  border-bottom: none;
  background: var(--header-background);
  gap: 1px;
}

.daysLarge,
.daysMedium,
.daysSmall {
  display: none;
}

.large .daysLarge,
.medium .daysMedium,
.small .daysSmall {
  display: grid;
}

.days .dayHeader {
  align-self: stretch;
  justify-self: stretch;
  background: var(--header-background);
  color: var(--header-color);
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 0.8em;
}

.grid {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  border: 1px solid var(--day-border-color);
  background: var(--day-border-color);
  gap: 1px;
}

.grid > div {
  display: grid;
  background: var(--calendar-background);
}

.grid > div > div {
  grid-area: 1 / 1 / 2 / 2;
}

.grid > div::before {
  content: '';
  padding-bottom: 100%;
  display: block;
  grid-area: 1 / 1 / 2 / 2;
}

.grid .placeholder {
  background: var(--placeholder-background);
}

.grid .day > div {
  padding: 0.5rem;
}

.small .grid .day > div {
  display: flex;
  align-items: center;
  justify-content: center;
}

.grid .day.today {
  background: var(--today-background);
}
</style>
    <div class=\\"calendar large\\">
      <div class=\\"header\\">
        <h2><span class=\\"month\\">June</span> <span class=\\"year\\">2020</span></h2>
      </div>
      <div class=\\"days daysLarge\\"><div class=\\"dayHeader\\">Sunday</div><div class=\\"dayHeader\\">Monday</div><div class=\\"dayHeader\\">Tuesday</div><div class=\\"dayHeader\\">Wednesday</div><div class=\\"dayHeader\\">Thursday</div><div class=\\"dayHeader\\">Friday</div><div class=\\"dayHeader\\">Saturday</div></div>
      <div class=\\"days daysMedium\\"><div class=\\"dayHeader\\">Sun</div><div class=\\"dayHeader\\">Mon</div><div class=\\"dayHeader\\">Tue</div><div class=\\"dayHeader\\">Wed</div><div class=\\"dayHeader\\">Thu</div><div class=\\"dayHeader\\">Fri</div><div class=\\"dayHeader\\">Sat</div></div>
      <div class=\\"days daysSmall\\"><div class=\\"dayHeader\\">S</div><div class=\\"dayHeader\\">M</div><div class=\\"dayHeader\\">T</div><div class=\\"dayHeader\\">W</div><div class=\\"dayHeader\\">T</div><div class=\\"dayHeader\\">F</div><div class=\\"dayHeader\\">S</div></div>
      <div class=\\"grid\\"><div class=\\"placeholder\\"></div><div class=\\"day\\"><div>1</div></div><div class=\\"day\\"><div>2</div></div><div class=\\"day\\"><div>3</div></div><div class=\\"day\\"><div>4</div></div><div class=\\"day\\"><div>5</div></div><div class=\\"day\\"><div>6</div></div><div class=\\"day\\"><div>7</div></div><div class=\\"day\\"><div>8</div></div><div class=\\"day\\"><div>9</div></div><div class=\\"day\\"><div>10</div></div><div class=\\"day\\"><div>11</div></div><div class=\\"day\\"><div>12</div></div><div class=\\"day\\"><div>13</div></div><div class=\\"day\\"><div>14</div></div><div class=\\"day\\"><div>15</div></div><div class=\\"day\\"><div>16</div></div><div class=\\"day\\"><div>17</div></div><div class=\\"day\\"><div>18</div></div><div class=\\"day\\"><div>19</div></div><div class=\\"day\\"><div>20</div></div><div class=\\"day\\"><div>21</div></div><div class=\\"day\\"><div>22</div></div><div class=\\"day\\"><div>23</div></div><div class=\\"day\\"><div>24</div></div><div class=\\"day\\"><div>25</div></div><div class=\\"day\\"><div>26</div></div><div class=\\"day\\"><div>27</div></div><div class=\\"day\\"><div>28</div></div><div class=\\"day\\"><div>29</div></div><div class=\\"day\\"><div>30</div></div><div class=\\"placeholder\\"></div><div class=\\"placeholder\\"></div><div class=\\"placeholder\\"></div><div class=\\"placeholder\\"></div></div>
    </div>
  "
`;
